### ISSUE TITLE
help implementing screensaver for ILI9341 tft lcd

### ISSUE THREAD

**Issue by ee-boi** [Pconti31/GUIslice-Solutions#5]
Hi @Pconti31, I am once again asking for help.

I was taking a look through your GUIslice-Solutions library and saw your example for a screensaver implementation for a TFT_eSPI display. I was wondering if you could help me to adapt this for use with an ILI9341 display with a tweak:

- use a bmp image (lets say my logo) that floats around the screen like an old DVD player screensaver ([example](https://www.bouncingdvdlogo.com/)), instead of the digital rain.
     - if that is too complicated, just moving the screen brightness to 0 would also be sufficient.

Any thoughts on how to do this? I must admit I looked through your example for a screen-saver app and am a bit confused on how I might go about this. Thinking I use your GUIslice_Image2C app to turn the logo to a bmp, store that in flash memory, but then how would I get it to bounce around the screen instead of the digital rain per the example?

**Specs:**
- [Adafruit Metro M0 express MCU (SAMD21, arm cortex M0+)](https://www.adafruit.com/product/3505)
- [ILI9341 display](https://www.adafruit.com/product/1770) 
- platformIO via VScode as environment
- GUIslice config selected: Custom (as previously posted in closed issue https://github.com/ImpulseAdventure/GUIslice/issues/509)

Any and all help is greatly appreciated. Thank you!

- ee-boi

**Comment by Pconti31** [comment]
@ee-boi I'm not sure myself since its been a while since the last time I looked at this code.
I'll take a look and see what I can come up with, may take a day or two.
Paul--

**Comment by ee-boi** [comment]
@Pconti31 very much appreciative of that. 

I did get your GUIslice_Image2C app to generate my logo using a jpg to a C array and tested via the ex28_ard_btn_img_flash example. So that part is working. (Although the easy part I suppose).

Thank you for your time!

**Comment by Pconti31** [comment]
@ee-boi Quick question what graphics driver are you using Adafruit_GFX or TFT_eSPI?

**Comment by ee-boi** [comment]
@Pconti31 using the Adafruit_GFX driver

**Comment by Pconti31** [comment]
@ee-boi Here you go
[screen-saver-logo.zip](https://github.com/Pconti31/GUIslice-Solutions/files/11909906/screen-saver-logo.zip)

I'll most likely post this in the apps section when I get some free time.
Feel free to ask and questions.

Paul--

**Comment by ee-boi** [comment]
@Pconti31 you are awesome. Going to work on getting this implemented in my project tonight and tomorrow morning. Will report back on progress.

Thank you so much again!

**Comment by ee-boi** [comment]
Hi @Pconti31 

Ok I had less time to spend on it over the past day then I thought. In any case, I did get the project uploaded, although I am seeing a bug where the countdown keeps resetting, even with no touch to the screen. Video examples attached.

Not sure what would cause this, have not had time to roll through the project proper yet to see, but wanted to post to see if you had any ideas.

I did switch the countdown to 1s and was able to get it to "kick-over" to the screensaver, the issue there being that it will not return back to the prior control screen.

I will spend sometime looking through over the weekend to see if I can find the source. Wondering if maybe it has something to do with my wiring or the touch pressure sensitivity?

[scrnsaver_bug_example.zip](https://github.com/Pconti31/GUIslice-Solutions/files/11918467/scrnsaver_bug_example.zip)

**Comment by Pconti31** [comment]
@ee-boi couple things 
1- you shouldn't use less than 2 seconds.  I guess you are trying it for debugging???

2- in your guislice config file uncomment
  //#define DBG_TOUCH         // Enable debugging of touch-presses

and check your serial monitor for output.  This should display all touch events.

3-   if you are using simple touch driver the touch sensitivity can be defined by the following parameters.
in your guislice config file.
ADATOUCH_PRESS_MIN
ADATOUCH_PRESS_MAX

Otherwise I'm afraid you are on your own...
Paul--

**Comment by ee-boi** [comment]
@Pconti31 

1. yes was just trying for debugging purposes, only to see if I could even get it to switch over.

2. will check out the debugger and report back

3. I did mess with the touch sensitivity a bit,  bringing the min touch value up from the default 10 to 500 to see if that did anything. Did not see a change.

I am also wondering if it has to do with the adafruit recommended wiring of the display, with X+ sharing the D/C pin.

In any case, I will muck around about and see what I figure out and report back. Thank you for the help.

![Screenshot_20230630_124334](https://github.com/Pconti31/GUIslice-Solutions/assets/125930238/be10d164-e364-4656-8e58-cc9f76a9056b)

**Comment by ee-boi** [comment]
@Pconti31 

Alright had a little time to work on this, still did not find a solution. I turned on debugging the touch presses as mentioned in comment 2 `//#define DBG_TOUCH // Enable debugging of touch-presses` and am seeing lots of "touch events" occurring even though in reality the screen is not being touched. Example of output below.

I tried the following:
- swap adafruit recommended wiring so X+ is on its own pin (not shared with D/C) -> no avail
- replaced display entirely with a spare in case original display is damaged -> no avail
- Adjust `ADATOUCH_PRESS_MIN` until serial monitor does not show touches -> value = 600
    -  when the value is at 600 for `ADATOUCH_PRESS_MIN` , it works as intended, albeit with the side effect that the resistance is too high to detect actual touch presses.
    - any value for `ADATOUCH_PRESS_MIN`  below 600 gives "ghost" touches. 

Any thoughts on what I could be missing here? Does it have something to do with `#define ADATOUCH_RX       300   // "rxplate"`?

Code snippet from Debug Serial Monitor:
```*** Timer! 10
DBG: Touch End  =0 Raw[2,1023] *****
DBG: remapX: (2,170,895,0,239)
DBG: remapY: (1023,95,895,0,319)
DBG: PreRotate: x=0 y=319
DBG: RotateCfg: remap=1 nSwapXY=1 nFlipX=0 nFlipY=1
DBG: Touch Press=0 Raw[2,1023] Out[319,239]
Trk: (319,239) P=0 : TouchUp

DBG: remapX: (1,170,895,0,239)
DBG: remapY: (1023,95,895,0,319)
DBG: PreRotate: x=0 y=319
DBG: RotateCfg: remap=1 nSwapXY=1 nFlipX=0 nFlipY=1
DBG: Touch Press=299 Raw[1,1023] Out[319,239]
Trk: (319,239) P=299 : TouchDown

DBG: Touch End  =0 Raw[1,1023] *****
DBG: remapX: (1,170,895,0,239)
DBG: remapY: (1023,95,895,0,319)
DBG: PreRotate: x=0 y=319
DBG: RotateCfg: remap=1 nSwapXY=1 nFlipX=0 nFlipY=1
DBG: Touch Press=0 Raw[1,1023] Out[319,239]
Trk: (319,239) P=0 : TouchUp

*** Timer! 10
DBG: remapX: (1,170,895,0,239)
DBG: remapY: (1023,95,895,0,319)
DBG: PreRotate: x=0 y=319
DBG: RotateCfg: remap=1 nSwapXY=1 nFlipX=0 nFlipY=1
DBG: Touch Press=299 Raw[1,1023] Out[319,239]
Trk: (319,239) P=299 : TouchDown

DBG: Touch End  =0 Raw[1,1023] *****
DBG: remapX: (1,170,895,0,239)
DBG: remapY: (1023,95,895,0,319)
DBG: PreRotate: x=0 y=319
DBG: RotateCfg: remap=1 nSwapXY=1 nFlipX=0 nFlipY=1
DBG: Touch Press=0 Raw[1,1023] Out[319,239]
Trk: (319,239) P=0 : TouchUp

*** Timer! 10
DBG: remapX: (2,170,895,0,239)
DBG: remapY: (1023,95,895,0,319)
DBG: PreRotate: x=0 y=319
DBG: RotateCfg: remap=1 nSwapXY=1 nFlipX=0 nFlipY=1
DBG: Touch Press=299 Raw[2,1023] Out[319,239]
Trk: (319,239) P=299 : TouchDown

DBG: Touch End  =0 Raw[2,1023] *****
DBG: remapX: (2,170,895,0,239)
DBG: remapY: (1023,95,895,0,319)
DBG: PreRotate: x=0 y=319
DBG: RotateCfg: remap=1 nSwapXY=1 nFlipX=0 nFlipY=1
DBG: Touch Press=0 Raw[2,1023] Out[319,239]
Trk: (319,239) P=0 : TouchUp

*** Timer! 10
DBG: remapX: (2,170,895,0,239)
DBG: remapY: (1023,95,895,0,319)
DBG: PreRotate: x=0 y=319
DBG: RotateCfg: remap=1 nSwapXY=1 nFlipX=0 nFlipY=1
DBG: Touch Press=299 Raw[2,1023] Out[319,239]
Trk: (319,239) P=299 : TouchDown

*** Timer! 10
DBG: Touch End  =0 Raw[2,1023] *****
DBG: remapX: (2,170,895,0,239)
DBG: remapY: (1023,95,895,0,319)
DBG: PreRotate: x=0 y=319
DBG: RotateCfg: remap=1 nSwapXY=1 nFlipX=0 nFlipY=1
DBG: Touch Press=0 Raw[2,1023] Out[319,239]
Trk: (319,239) P=0 : TouchUp

DBG: remapX: (1,170,895,0,239)
DBG: remapY: (1023,95,895,0,319)
DBG: PreRotate: x=0 y=319
DBG: RotateCfg: remap=1 nSwapXY=1 nFlipX=0 nFlipY=1
DBG: Touch Press=299 Raw[1,1023] Out[319,239]
Trk: (319,239) P=299 : TouchDown

*** Timer! 10
DBG: Touch End  =0 Raw[1,1023] *****
DBG: remapX: (1,170,895,0,239)
DBG: remapY: (1023,95,895,0,319)
DBG: PreRotate: x=0 y=319
DBG: RotateCfg: remap=1 nSwapXY=1 nFlipX=0 nFlipY=1
DBG: Touch Press=0 Raw[1,1023] Out[319,239]
Trk: (319,239) P=0 : TouchUp

DBG: remapX: (1,170,895,0,239)
DBG: remapY: (1023,95,895,0,319)
DBG: PreRotate: x=0 y=319
DBG: RotateCfg: remap=1 nSwapXY=1 nFlipX=0 nFlipY=1
DBG: Touch Press=299 Raw[1,1023] Out[319,239]
Trk: (319,239) P=299 : TouchDown

DBG: Touch End  =0 Raw[1,1023] *****
DBG: remapX: (1,170,895,0,239)
DBG: remapY: (1023,95,895,0,319)
DBG: PreRotate: x=0 y=319
DBG: RotateCfg: remap=1 nSwapXY=1 nFlipX=0 nFlipY=1
DBG: Touch Press=0 Raw[1,1023] Out[319,239]
Trk: (319,239) P=0 : TouchUp

*** Timer! 09
DBG: remapX: (1,170,895,0,239)
DBG: remapY: (1023,95,895,0,319)
DBG: PreRotate: x=0 y=319
DBG: RotateCfg: remap=1 nSwapXY=1 nFlipX=0 nFlipY=1
DBG: Touch Press=299 Raw[1,1023] Out[319,239]
Trk: (319,239) P=299 : TouchDown

*** Timer! 10
DBG: Touch End  =0 Raw[1,1023] *****
DBG: remapX: (1,170,895,0,239)
DBG: remapY: (1023,95,895,0,319)
DBG: PreRotate: x=0 y=319
DBG: RotateCfg: remap=1 nSwapXY=1 nFlipX=0 nFlipY=1
DBG: Touch Press=0 Raw[1,1023] Out[319,239]
Trk: (319,239) P=0 : TouchUp

DBG: remapX: (1,170,895,0,239)
DBG: remapY: (1023,95,895,0,319)
DBG: PreRotate: x=0 y=319
DBG: RotateCfg: remap=1 nSwapXY=1 nFlipX=0 nFlipY=1
DBG: Touch Press=299 Raw[1,1023] Out[319,239]
Trk: (319,239) P=299 : TouchDown

DBG: Touch End  =0 Raw[1,1023] *****
DBG: remapX: (1,170,895,0,239)
DBG: remapY: (1023,95,895,0,319)
DBG: PreRotate: x=0 y=319
DBG: RotateCfg: remap=1 nSwapXY=1 nFlipX=0 nFlipY=1
DBG: Touch Press=0 Raw[1,1023] Out[319,239]
Trk: (319,239) P=0 : TouchUp

*** Timer! 09
DBG: remapX: (2,170,895,0,239)
DBG: remapY: (1023,95,895,0,319)
DBG: PreRotate: x=0 y=319
DBG: RotateCfg: remap=1 nSwapXY=1 nFlipX=0 nFlipY=1
DBG: Touch Press=598 Raw[2,1023] Out[319,239]
Trk: (319,239) P=598 : TouchDown

DBG: Touch End  =0 Raw[2,1023] *****
DBG: remapX: (2,170,895,0,239)
DBG: remapY: (1023,95,895,0,319)
DBG: PreRotate: x=0 y=319
DBG: RotateCfg: remap=1 nSwapXY=1 nFlipX=0 nFlipY=1
DBG: Touch Press=0 Raw[2,1023] Out[319,239]
Trk: (319,239) P=0 : TouchUp```

**Comment by Pconti31** [comment]
I see all the fake touches are x 319 y 239 a clue I guess. Changing adafruit_rx to higher values might help.

**Comment by Pconti31** [comment]
@ee-boi I seem to remember you are using a UNO R4 a 32 bit  Cortex-M4.   I think that version has a 12 bit ADC not a 10 bit one like the AVR's have.  If so this could be screwing up reading from the A2 and A3 pins giving a false reading. 
There is a Arduino routine analogReadResolution which can your board to use ten bits like so analogReadResolution((10);
You can edit GUIslice_drv_adagfx.cpp commenting out the check for ESP32 lines starting at 2196
```
  // For ESP32 devices, a workaround is required for the
  // Adafruit_TouchScreen since it makes an assumption that
  // the ADC resolution is 10-bit. This workaround enables the
  // Adafruit library to operate the same was as for AVR devices.
  //#if defined(ESP32)
    #define FIX_4WIRE_ADC_10 // Comment out to disable
  //#endif
```
If that doesn't work try playing commenting out one or both in the same area.
```
  #define FIX_4WIRE_Z // Comment out to disable
  #define FIX_4WIRE_PIN_STATE // Comment out to disable
```


Maybe that will help?
If all of that fails I can look at modifying Adafruit_TouchScreen/TouchScreen.cpp to add a time delay between readings or if that fails skip any reading of x=319, y-239.
Paul--

**Comment by ee-boi** [comment]
@Pconti31 

Something very strange going on…it is indeed always showing a touch at x319 y239, I can actually see the little yellow mark from your code in the corner there. 

I tried a whole bunch of things - swapped the wire, commented out code per your previous post, tried another resistive touchscreen I had as a spare. All showing same issue.

Then I swapped the X+ pin from A5 to A1 and voila, it worked fine. But without finding root cause I meddled around with the wiring and continued to switch pins and then the same issue returned again. Truly perplexed by this, all I can think right now is there is some electrical or connection issue somewhere…

Going to keep troubleshooting. Probably something I am missing. 

Also you are correct - I am using the metro m0 express board (arm cortex m0+), which has a 12-bit ADC.

**Comment by Pconti31** [comment]
@ee-boi Maybe a pin is floating.  Try attaching unused digital and analog pins to ground.  
Also make jumper wires as short as possible so they don't act as an antenna.

**Comment by ee-boi** [comment]
Ok - after a lot of troubleshooting, I believe I have fixed the issue. Now, I am not certain that I addressed the exact root cause but I no longer seem to have a phantom touch at X329 Y239 any more.

What I did (that worked):
1. grounded the only other analog pin I was not using on the dev board (A5)
2. edited the Adafruit_touchscreen library .cpp file, replacing entries with 1023 -> 4095 and 1024 -> 4096, effectively switching it from 10 bits to 12 bits
3. adding in `analogReadResolution(12); `in `setup();` function on main sketch
4. re-calibrating using diag_ard_touch_calib example and updating my custom config with those values

After this the screen-saver-logo example you sent seemed to work fine, no longer a phantom touch at the corner. I also noted that when dragging my stylus around it seemed slightly more accurate as far as positioning the press to where the actual touch was.

Anyway - I was able to get my logo into a C array using your GUIslice Image2C program, and bouncing around like the example.

Now one more (hopefully last) question for you, is there anyway to modify the BouncingLogoAnim code so that the inactivity timeout can be changed at runtime by a user and not in the initialization of the object?

**Comment by ee-boi** [comment]
@Pconti31 

Okay hopefully I am not spamming you. I was able to get the screensaver logo adapted into my main project and here are the questions/issues that have arisen:

1. per previous post, can the bouncinglogoanim or screenmanager code be modified so timeout can be modified at run time by a user?

2. when the screensaver activates, and then I touch the display to pause it (touch event occured) it does not return to any screen, just gives me a blank display and then after 10s it times out back to the screensaver again. Am I missing a call to return back to previous page somewhere I need to adjust for my project?

3. Should the buttons that are behind the screensaver be disabled so they can't erroneoulsy be pressed when "waking" the display out of screensaver mode? Or is there a better way to do this like a large invisible popup or something? I ask because the buttons in my project move a stepper motor and I would like to avoid that if possible when trying to "wake" from the screensaver

4. I put all of the files (both .cpp & .h) (BouncingLogoAnim, ScreenManager, logo.c, gslc_touchAny) into the /src directory on platformio. It works fine, but shouldn't these go into /include? When I tried /include it did not compile.

**Comment by Pconti31** [comment]
@ee-boi Happy you found your problem with your floating pin.  
1.  I modified my sample app ScreenManager now has a runtime call setTimeout(seconds);  Check out lines 148-152 of screen-saver-logo.ino.  It now has optional input for changing timeout.  Note that you might want to issue a gslc_PopupHide(&m_gui); inside pause() just in case the user is inside a keyboard when the timeout occurs.  Refresh of the screen might break otherwise.
2.  I don't understand why this would happen.  My original screen saver had more than one page and always showed the current page after a pause().  If you can post a zip file of an example I'll check it out.
3. Disabling these buttons might be a good idea.  While it shouldn't be necessary it always good to be safe.  It's called defensive programming.
4. I would think that you would place all *.h files in include and all *.c and *.cpp in src.
Paul--

**Comment by ee-boi** [comment]
@Pconti31 

Appreciate your responses. I was able to get the runtime call setTimeout() working! Thank you so much!

Ok as for number 2, I unfortunately do not want to post my code in public on Github as it ultimately will not be mine (it is for a friend/clients project). It also is fairly spaghetti so there would be a lot of wasted info in there you need to sort through. I will work on making an example of something similar to post.

That said below is a video of what is happening when I try to "wake" the display from the screen saver. I have noted once in awhile when I do this a single button or text will appear on the blank screen, but the full page never returns.

I have many elements stored in flash memory so as to keep RAM open for usage by the function calls of the program and whatnot. Would this maybe have some effect on it? I also renamed pages like "E_PG_SETUP, E_PG_MODE1", etc.

Otherwise all else is the same as your example as far as I can tell (dumped files into /src, added lines for ScreenManager, BouncingLogoAnim and the logo array at top of file,  added `myAnimation.init()` & `myAnimation.pause()` & `screenMgr.init() `into `setup()`, replaced `gslc_Update()` with `screenMgr.update()` in` loop()` )

[blankscreen.zip](https://github.com/Pconti31/GUIslice-Solutions/files/11963016/blankscreen.zip)
[blankscreen2.zip](https://github.com/Pconti31/GUIslice-Solutions/files/11963067/blankscreen2.zip)

**Comment by Pconti31** [comment]
@ee-boi 
 I looked at the videos and it's not really any help.   I need to be able to set breakpoints or place trace statements.   

Since you have spaghetti code make sure there is only one call to screenMgr.update(); and absolutely no calls to gslc_Update(&m_gui); If necessary, create your own update function that all of code calls and place  screenMgr.update(); in that new routine.

I'll modify my sample to add a few extra pages and see what happens.  

Otherwise without code to follow I really can't be of any further help.

Sorry!
Paul--

**Comment by ee-boi** [comment]
@Pconti31 

Ok no problem. I should have time in the coming days and will work on drafting a similar example that uses 4 pages and a popup, with most items stored in flash. 

Will post it soon.

Appreciate your help Paul.

Thank you!

**Comment by Pconti31** [comment]
@ee-boi I made some enhancements to gslc_touchAny() to eat all touch events when it detects the first one.  this will make it more responsive and you won't need to disable any buttons.  The test prohram now has a second page filled with buttons.  Button labeled 1 to 3 will jump back to  page main.  You can comment out the MY_DEBUG at the top of gslc_touchAny.c to avoid touch tracing to serial output.
Paul--

**Comment by ee-boi** [comment]
@Pconti31 

Ok I made an example using the builder with very similar buttons and pages to what I am using. Upon testing it has the same issue as the previous video - after ten seconds the screensaver kicks in, then a press on the screensaver "wakes" it to just a black screen & not back to the previous page, then after 10 seconds the screensaver starts again.

I will also note - this is using your most up to date code from the screensaver-logo repo, the updates actually made touch on the regular pages less responsive for some reason also... it is strange after several presses a button will highlight but wont do anything? ...but the screensaver does register touch events, it just goes to a black screen as mentioned.

Attached is a zip of the example, I have also put my custom config and GUIslice_drv_adagfx.cpp file in as well.

[screensavertest3.zip](https://github.com/Pconti31/GUIslice-Solutions/files/11992275/screensavertest3.zip)
[configs.zip](https://github.com/Pconti31/GUIslice-Solutions/files/11992277/configs.zip)

I am using the adafruit metro M0 express dev board and the adafruit 2.8" tft resistive touchscreen (ILI9341)

**Comment by Pconti31** [comment]
@ee-boi I got your sample running with my latest bug fixes to ScreenManager.(h cpp).  The only change to gslc_touchAny.c was to comment out #define MY_DEBUG

Paul--

**Comment by Pconti31** [comment]
@ee-boi One other point you have a bug on page3 that will cause you to hang if you select the popup.
you have a bug on lines 151-153
```
      case E_ELEM_BTN13:
        gslc_SetPageCur(&m_gui, E_PG_POPUP1RUN);
        break;
```
should be:

```
      case E_ELEM_BTN13:
        gslc_PopupShow(&m_gui, E_PG_POPUP1RUN, true);
        break;
```
 Paul--

**Comment by ee-boi** [comment]
@Pconti31 

do you mean comment out #define MY_TOUCH ?

Also - so the example I sent over worked fine for you? It returned back to the previous screen when exiting the screensaver?

EDIT: here is a video of the exact same code I sent over running on my end :(
[issues3.zip](https://github.com/Pconti31/GUIslice-Solutions/files/11995639/issues3.zip)

**Comment by Pconti31** [comment]
@ee-boi Yes, with my newly checked in ScreenManager.(h cpp) and with the bug fix to screensavertest3.ino your sample  works fine now.  You did pull down the new code?

**Comment by ee-boi** [comment]
@Pconti31

huh. Strange. I did, let me try again.

**Comment by ee-boi** [comment]
@Pconti31 

Ok I made sure I pulled down the new code and copied over the files into the example project folder, leaving the .ino and *_GSLC.h alone that were generated by the builder.

**Copied over files from github repo:**
- BouncingLogoAnim.cpp  & .h
- ScreenManager.cpp  & .h
- dvd-logo.c
- gslc_touchAny.c  & .h

Now, it does return back to the previous page when touched during the screensaver being on (yay!), but still has an issue detecting touch events on the screen. 

Another video attached to show what the issue is. (note: I uncommented `#define DBG_TOUCH` in my custom `ard-adagfx-ili9341-simple-m0express.h` file to ensure that touches were registering in the correct location, they are).

[issues4.zip](https://github.com/Pconti31/GUIslice-Solutions/files/11995942/issues4.zip)

**Comment by Pconti31** [comment]
@ee-boi 
try setting in your config file
```
  #define GSLC_TOUCH_MAX_EVT    30
```
and see if that works.
Otherwise you will have to debug this on your own checking the trace output and adding any additional trace statements you need.
uncomment #define MY_DEBUG and look at your output.  
Since It works fine for me I can't really be of much help.   
Paul--

**Comment by ee-boi** [comment]
@Pconti31 

changing  `#define GSLC_TOUCH_MAX_EVT    30` certainly helps, but it is still not as responsive as when I do not have any of the screensaver files included at all.

Also - I don't see `#define MY_DEBUG` anywhere, all I see is `#define MY_TOUCH` in the updated `gslc_touchAny files`, is this what you are referring to?

So I tried something else, I took the latest project file I sent you and swapped the old `gslc_touchAny.c` and the `gslc_touchAny.h` from before they were altered with the `my_TrackTouch` & `my_touch_loop` functions and loaded that. The screen is responsive now like it was previously so my issue for touch response lies in that file somewhere.

I'll give it a little more time trying to debug further, if not I will just have the display backlight be on a timer and go to 0 and call it good.

Edit: Ok my two issues lie in the following

1. **touch lack of response:** Issue occurs when using the updated `gslc_touchAny `files. When I swap back to the older files the issue goes away.

2. _**_Does not return to previous page when trying to wake from screensaver:_**_ Only occurs when using
 ``` 
void my_display() { 
// myAnimation doesn't really loop
  // but just paints one frame then returns
  myAnimation.loop();
} ```

**Comment by Pconti31** [comment]
@ee-boi yes, I meant MY_TOUCH.
Well, I can take a hint.  I'll investigate gslc_touchAny.c further and see if I can figure out why its not responding and if not I'll go back to the original version.  Maybe, I could also get ScreenManager to make one large full screen button and use that for restore than the old version of gslc_touchAny.c  won't be involved except  for resetting timeouts thus preventing buttons underneath from getting triggered.  Then you wouldn't need to worry about your stepper motors firing up when some restores a from screen saver mode.

Changing  gslc_touchAny.c  won't effect screen restore, only swapping out the older screen manager should do that.  
The new screen manager hides any active popup then when a touch event occurs it rips through the currently active page and sets each ui element on it to needing redraw which you should see inside ScreenManager::update() function.

Paul--

**Comment by ee-boi** [comment]
@Pconti31 

No aggressiveness intended in that last message, more so frustration. I feel it should be working on my end and do not understand what is going wrong.

In any case, appreciate you looking into it gslc_touchAny.c it does work fine for me with the older file.

As far as not being able to return to the current page when trying to wake from home screen (issue 2 in previous comment), I retract that it is is only occurring when using 
``` 
void my_display() { 
// myAnimation doesn't really loop
 // but just paints one frame then returns
 myAnimation.loop();
} 
``` 
I now think it has something to do with having multiple pages? 

At this point in the troubleshooting I am puzzled. I am not sure why some examples work and others do not, when the files are the same. It's almost like when flashing the board it flashes the code slightly different per example, I don't get it...

**Comment by ee-boi** [comment]
@Pconti31 

Sending over another video and a zip of the complementing code. It is quite strange, when attempting to "wake" from the screensaver if I touch where a button or other element is it will refresh through the blank screen. and if I touch where a button that jumps to a new page is it will refresh the screen, but if you notice at the 1:10 mark in the video it does not return to the full screen resolution...??

Something to do with the screenManager lines 65-67 maybe?

[issues5.zip](https://github.com/Pconti31/GUIslice-Solutions/files/12003406/issues5.zip)
[screensavertest5.zip](https://github.com/Pconti31/GUIslice-Solutions/files/12003410/screensavertest5.zip)

Note: this code is using the files from the older scrensaver-logo repo, and my own example *_GSLC.h from the builder.

**Comment by Pconti31** [comment]
@ee-boi Well of course its not working.  You must use the new version of ScreenManager its the only one that will refresh your page.

**Comment by ee-boi** [comment]
@Pconti31 

you are correct, I must have got them mixed up as I was trying different things and debugging. Apologies for any added confusion.

OK - The attached code works! I am using the most up to date BouncingLogoAnim .cpp/.h, & ScreenManager .cpp/.h files along with the older previous gslc_touchAny .c/.h files. Screensaver starts after ten seconds and upon touch event(s) does wake and return to the previous page.

The only more minor issue(s) I have now are sometimes it takes several presses (or a hold) to wake from the screensaver and I am able to accidentally press buttons "behind" the screensaver when waking. 

Is it better I disable the buttons as we previously discussed? Or if I haven't exhausted your good will yet, are you still taking a crack at modifying the gslc_touchAny files to eat all touch events? 

I do really appreciate all your help on this, and apologies for my MechE brain. I am not sure I can offer you anything, but if I could I would.

[screensavertest6.zip](https://github.com/Pconti31/GUIslice-Solutions/files/12010057/screensavertest6.zip)

**Comment by Pconti31** [comment]
@ee-boi I have invented a new scheme that i hope will work better.
The idea is to use mt gslc_touchAny() only to detect non-activity so we can trigger a timeout. After that when the logo image is displaying guislice handles all touch event and screen refresh.   Seems to be working nicely with your screensavertest5 so I have attached a zip file with all of the changes.  If it works for you I will check it in to my repro.
[screensavertest5.zip](https://github.com/Pconti31/GUIslice-Solutions/files/12016145/screensavertest5.zip)

Theory behind the new version:
During ScreenManger init() I create a hidden page and place a box with touch and draw callbacks on it.  When we detect non-activity that triggers screen saver drawing I bring this hidden page up as a popup. (I tried making a normal page but base page continued to draw on screen!).  This new hidden popup being modal means you don't need to fear your stepper motors firing up accidently.
Paul--

**Comment by ee-boi** [comment]
@Pconti31 

**Paul - just tested...everything appears to be working great! THANK YOU SO MUCH!!** :)

I do have two follow up questions (but this should be getting to the end of it):

1. I added `gslc_PopupHide(&m_gui);` into `pause() `in the `*.ino` file per one of your earlier recommendations. Is this still the recommended action? _(the reason I ask is - I have an e-stop button that kills all power to the motor and displays a popup page, the way to restore power to the motor is by pushing a button available on the e-stop popup page, that also then hides the estop popup page. Thinking if the screensaver is activated while the estop popup page is active and then hides the estop popup per the call in `pause()`, the user will have to press the e-stop button again, get the popup back, then clear it with the button on the popup page. A relatively minor inconvenience, but an extra step for the end user)_

2. is there a way I can stop the screensaver from loading indefinitely? I would like to have something like a simple checkbox the user can click that prevents the screensaver from turning on at all if they so choose.

I will start migrating this code over to my main project and report back as well. Want to double down on my appreciation for all your help again, thank you.

**Comment by Pconti31** [comment]
@ee-boi 
1.  No.  I close any open popups so i can do my invisible popup.
2. I Suppose you could add a global boolean and if it false not call   screenMgr.update();
```
  if (g_bUseScreenSaver)
       screenMgr.update();
  else
     gslc_Update(&m_gui);
```
Please leave a star in this repository  if you are satisfied.
Paul--

**Comment by Pconti31** [comment]
@ee-boi As I think more about item 2 above the global boolean won't work.  The timer will get screwed up.
I'll add a disable() call to ScreenManger before doing my check in.
Paul--

**Comment by Pconti31** [comment]
@ee-boi Everything is check into my repro.  

New calls ScreenManager::disable() and ScreenManager::enable() to restore screen saver after doing a disable();

Again please leave a star, if you are satisfied.
Paul--

**Comment by ee-boi** [comment]
@Pconti31 

Getting tied up on another project at the moment, should have time soon to get the changes implemented on my end. So far, everything seems to be working as intended!!

Can’t tell you how much I appreciate your help Paul, I would have been up a creek without you or GUIslice (or the builder) for that matter.

Thank you! As always will report back once I get it implemented into my project!

**Comment by ee-boi** [comment]
@Pconti31 

Ok so I got this fully integrated and all works well. The animation in the polling loop does affect movement of my motor when the screensaver is on, I believe this to be an issue more with Accelstepper than with the GUIslice code. Going to work on trying to create a pulse signal using interrupts and timers for now. 

In any case , this work fine and I want to thank you Paul for you time and diligence helping me get to a solution. I dropped a star on the repo. Closing this for now. Thank you!!!!!! :)
